core/domain/stats_jobs_one_off.py

```diff
@@ -73,6 +73,104 @@ def reduce(exp_id, values):
         yield "%s: %s" % (exp_id, values)
 
 
+class RegenerateMissingStatsModelsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off job to regenerate missing stats models for explorations with
+    correct v1 stats from previous versions.
+
+    Note that job RecomputeStatisticsOneOffJob must be run immediately after
+    this one, otherwise the v2 stats will be inconsistent.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(exploration_model):
+        exp_stats_instances = []
+
+        # Find the latest version number.
+        exploration = exp_services.get_exploration_by_id(exploration_model.id)
+        latest_exp_version = exploration.version
+        versions = range(1, latest_exp_version + 1)
+
+        # Get list of snapshot models for each version of the exploration.
+        snapshots_by_version = (
+            exp_models.ExplorationModel.get_snapshots_metadata(
+                exploration.id, versions))
+        if not exploration_model.deleted:
+            current_version = exploration_model.version
+            for exp_version in xrange(1, current_version + 1):
+                exp_stats_model = (
+                    stats_models.ExplorationStatsModel.get_model(
+                        exploration.id, exp_version))
+                if not exp_stats_model:
+                    state_stats_mapping = {
+                        state_name: stats_domain.StateStats.create_default()
+                        for state_name in exploration.states
```

Comment:
This seems wrong, shouldn't it use the state names from the exploration version in question? exploration.states is the latest version.

Or alternatively you can generate it by looking at the previous state stats mapping and modifying it according to additions/deletions/renames.

Reply:
Nice catch. Fixed by doing a get_multi_versions for all versions of the exploration at the outset and using the particular version of the exploration.states where required.